### PR TITLE
First step to TypeScript upgrade

### PR DESCRIPTION
**Pull request by JustinBeckwith**
First steps towards #503.  This simply renames all the files (including generated ones) to *.ts, and makes any changes necessary to get the build and tests to pass.  I am going to update #503 with the full list of steps we need to take here.  

- [x] `npm test` succeeds
- [x] Pull request has been squashed into 1 commit
- [ ] I did NOT manually make changes to anything in `apis/` 
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate JSDoc comments were updated in source code (if applicable)
- [x] Appropriate changes to readme/docs are included in PR

I chose to hand modify apis/ to mollify the TypeScript compiler, and because I didn't want to complicate the PR with schema changes.

### CODE REVIEW

**Comment by JustinBeckwith** [comment]
@matthewloring since @ofrobots is out, can you take a look?

File: package.json [hunk 1]
```diff
@@ -77,16 +79,20 @@
...
-    "swig": "1.4.2"
+    "source-map-support": "^0.4.14",
+    "swig": "1.4.2",
+    "typescript": "^2.2.2"
   },
   "scripts": {
-    "lint": "semistandard \"**/*.js\"",
-    "generate-apis": "node scripts/generate",
+    "lint": "semistandard \"samples/**/*.js\"",
+    "generate-apis": "npm run build-tools && node scripts/generate.js",
```
Comment (jmdobry) [inline package.json]: Shouldn't `build-tools` run _after_ `scripts/generate.js`?

Comment (JustinBeckwith) [inline package.json]: Comically, no.  build-tools actually builds generate.ts into generate.js so it can be run by node.

Comment (jmdobry) [inline package.json]: does generate.js not produce more .ts files?

**Comment by lukesneeringer** [comment]
> This simply renames all the files (including generated ones) to *.ts

This seems wrong to me unless the thing that generates the code has been updated to spit out `*.ts` files. Hand-edits to autogen code need to not happen.

File: package.json (hunk 1 above)
Comment (JustinBeckwith) [inline package.json]: Yes :) `npm run build-tools` generates `generate.js`, and `npm run generate-apis` generates `apis/*.ts`, which `npm run build` compiles into js.  Sorry - I know this is a little confusing.

**Comment by JustinBeckwith** [comment]
@lukesneeringer I updated the generators and templates to emit *.ts files, re-ran the generator, and ensured that it all worked.  The problem is that regenerating the ts files is non-deterministic.  The PR would include surface changes, and about 200k lines across all generated files to review.

It's really up to you guys.  I'm trying to limit the scope and improve the readability of the PR - but I'm happy to throw a watermelon over the fence :)

**Comment by lukesneeringer** [comment]
If you updated the generator, that resolves my complaint.

File: lib/discovery.ts [hunk 2]
```diff
@@ -59,13 +58,14 @@ function makeMethod (schema, method, context) {
       params: params,
       requiredParams: method.parameterOrder || [],
       pathParams: getPathParams(method.parameters),
-      context: context
+      context: context,
+      mediaUrl: null
```
Comment (matthewloring) [inline lib/discovery.ts]: Why is this option added? If this was done only for typescript maybe it should be made optional at the definition site.

File: test/test.clients.ts [hunk 3]
```diff
@@ -131,10 +133,10 @@ describe('Clients', function () {
   it('should support default params', function (done) {
     var google = new googleapis.GoogleApis();
     var datastore = google.datastore({
-      version: 'v1beta2',
+      version: 'v1beta3',
```
Comment (matthewloring) [inline test/test.clients.ts]: Why were these logical changes included as well?

File: lib/discovery.ts (hunk 2 above)
Comment (JustinBeckwith) [inline lib/discovery.ts]: It was done just for TypeScript, but this object literal is what formed the implicit type the compiler was checking against.  When I do a proper typing PR, I will move this to an interface with an optional parameter.

File: test/test.clients.ts (hunk 3 above)
Comment (JustinBeckwith) [inline test/test.clients.ts]: Funny story!  The v1beta2 release of the datastore *doesn't exist anymore*.  These unit tests are currently failing in the master branch.  I switched to `v1beta3` which actually changed the API surface.  So ...  I had to switch the unit tests.

Comment (matthewloring) [inline test/test.clients.ts]: It feels like this change should be separated from the move to typescript but the owners of the repository would probably be the right people to comment on whether this is ok.

Comment (jmdobry) [inline test/test.clients.ts]: keep this change.

**Comment by jmdobry** [comment]
I'm going to test this branch against our Node.js samples that use google-api-nodejs-client

**Comment by jmdobry** [comment]
Tested against our googleapis samples in the nodejs-docs-samples repo.

**Pull request merged**